### ISSUE TITLE
Patch logic fails in presence of two kustomize.Directory's

### ISSUE THREAD

**Issue by ossareh** [pulumi/pulumi-kubernetes#2437]
### What happened?

Using `@pulumi/kubernetes/kustomize/Directory` twice in a pulumi module results in an incorrect merge logic for patches.

### Expected Behavior

Regardless of the number of calls to `new k8s.kustomize.Directory()` in a module, the appropriate patch semantics should be honored.

### Steps to reproduce

With this directory structure:
* `internal_tools/`
  * `index.ts`
  * `tool_1/`
    * `base/`
      * `kustomize.yaml`
      * `deployment.yaml`
     * `overlays/production/`
       * `kustomize.yaml`
       * `patches/deployment.yaml # <- strategic merge`
  * `tool_2/`
    * `base/`
      * `kustomize.yaml`
      * `deployment.yaml`
     * `overlays/production/`
       * `kustomize.yaml`
       * `patches/deployment.yaml # <- strategic merge`

If index.ts looks something like this:

```typescript
import * as k8s form "@pulumi/kubernetes";

new k8s.kustomize.Directory(`tool_1`, {
  directory: `tool_1/overlays/production`
});

new k8s.kustomize.Directory(`tool_2`, {
  directory: `tool_2/overlays/production`
});
```

`tool_1` has it's patch applied incorrectly. Instead of it having the `environment` key merged it is replaced; i.e. `tool_1/base/deployment.yaml` values are lost.

If index.ts looks something like this:

```typescript
import * as k8s form "@pulumi/kubernetes";

new k8s.kustomize.Directory(`tool_1`, {
  directory: `tool_1/overlays/production`
});
```

`tool_1` applies correctly.

### Output of `pulumi about`

```
CLI
Version      3.69.0
Go Version   go1.20.4
Go Compiler  gc

Plugins
NAME    VERSION
nodejs  unknown

Host
OS       ubuntu
Version  20.04
Arch     x86_64

This project is written in nodejs: executable='/home/coder/.local/share/rtx/installs/nodejs/18.15.0/bin/node' version='v18.15.0'

Backend
Name           pulumi.com
URL            https://app.pulumi.com/ossareh
User           ossareh
Organizations  ossareh, erisyon

Pulumi locates its logs in /tmp by default
warning: Failed to get information about the Pulumi program's dependencies: could not find either /home/coder/erisyon/ops/packages/services/internal/yarn.lock or /home/coder/erisyon/ops/packages/services/internal/package-lock.json
warning: Failed to get information about the current stack: No current stack
```

The warning at the end is because I'm in an npm managed monorepo using workspaces where `package-lock.json` is in the root.

### Additional context

Using `kubectl apply -k` directly on each directory works without issue.

Adding an `internal_tools/kustomization.yaml` which has:

```yaml
resources:
  - tool_1
  - tool_2
```

And an index.ts like this:

```typescript
import * as k8s form "@pulumi/kubernetes";

new k8s.kustomize.Directory("all_tools", {
  directory: "."
});
```

Works just fine also.

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

**Comment by rquitales** [comment]
Hi @ossareh, thank you for reporting this issue and for the thorough analysis you provided, including the potential workaround and error modes. It seems that the problem lies in the configuration and creation of our Kustomize runner.

To address this, I will proceed by creating a test case to reproduce the issue and further investigate a potential solution for this. Thanks for bringing this to our attention!

**Comment by mjeffryes** [comment]
Unfortunately this issue has sat idle for a while; do you still observe this problem in the latest version of the provider?

**Comment by ossareh** [comment]
@mjeffryes I'm no longer at the company that was experiencing this issue. The prior comment suggested a test case would be written to verify the issue; did that happen?

**Comment by mjeffryes** [comment]
no, I don't think we have enough detail in the issue to fully reproduce as it looks like the behavior depends on the specific contents of the kustomize directories. (Specifically, I don't know how to produce the specific patches that produced the error.)

**Comment by ossareh** [comment]
Reading back my comments; I'm not entirely sure why I didn't provide an example, but maybe it's because it worked for trivial cases, something like setting `a = 1`.

**Comment by mjeffryes** [comment]
Since we don't have a solid reproduction at this time, I'm going to close this issue for now, but feel free to open a new issue if this problem shows up again!
